File: src/precedent/analysis/identity.py
"""Deciding when two award records are the same organization.

This is the most consequential code in the project. The repeat-winner and new-entrant rates
are the two statistics people will quote, and both are counts of *distinct organizations*.
Split one grantee into two and a program looks more open than it is; merge two into one and
it looks closed. Either way the number reads as authoritative.

Resolution runs in descending order of how much the identifier is worth:

1. **Unique Entity Identifier.** Assigned, unique, and the reason SAM.gov exists.
2. **USAspending's recipient id**, with its level suffix stripped. The same organization
   appears as ``...-C`` (child), ``-R`` (recipient) and ``-P`` (parent); keeping the suffix
   splits one organization into three.
3. **Normalized name.** A last resort, and the report says how often it was needed, because
   a program resolved mostly by name has weaker statistics than one resolved by UEI.
"""

from __future__ import annotations

import re
import unicodedata
from collections import Counter
from dataclasses import dataclass
from typing import Protocol

# Applied repeatedly: "FOO SERVICES INC LLC" loses both, "THE FOO CO" loses CO then THE.
LEGAL_SUFFIXES = (
    "INCORPORATED",
    "CORPORATION",
    "COMPANY",
    "LIMITED",
    "INC",
    "LLC",
    "LLP",
    "LP",
    "CORP",
    "CO",
    "LTD",
    "PC",
    "PA",
    "THE",
)

# Whole-token expansions only. Substring replacement would turn "COSTA" into "COMPANYSTA".
ABBREVIATIONS = {
    "DEPT": "DEPARTMENT",
    "UNIV": "UNIVERSITY",
    "ASSN": "ASSOCIATION",
    "ASSOC": "ASSOCIATION",
    "NATL": "NATIONAL",
    "INTL": "INTERNATIONAL",
    "SVCS": "SERVICES",
    "SVC": "SERVICES",
    "CTR": "CENTER",
    "CNTY": "COUNTY",
    "US": "UNITED STATES",
    "AAA": "AREA AGENCY ON AGING",
}

# "ST LOUIS" is a city and "ST STATE" is nonsense, so this expands only in first position,
# where "ST DEPARTMENT OF HEALTH" means the state's.
FIRST_TOKEN_ABBREVIATIONS = {"ST": "STATE"}

_NOT_ALNUM = re.compile(r"[^A-Z0-9 ]")
_SPACES = re.compile(r"\s+")
_LEVEL_SUFFIX = re.compile(r"-(C|R|P)$", re.IGNORECASE)

# The share of records resolved by name above which the statistics deserve a warning.
NAME_TIER_WARNING_THRESHOLD = 0.10


def normalize_name(raw: str | None) -> str:
    """One organization name reduced to a comparable form.

    Used for identity here and for pass-through name clustering, deliberately the same
    function: two places that normalize differently would disagree about the same
    organization, and that disagreement would be invisible.
    """
    if not raw:
        return ""
    text = unicodedata.normalize("NFKD", raw).upper()
    text = "".join(c for c in text if not unicodedata.combining(c))
    text = text.replace("&", " AND ")
    text = _NOT_ALNUM.sub(" ", text)
    text = _SPACES.sub(" ", text).strip()
    if not text:
        return ""

    tokens = text.split()
    if tokens and tokens[0] in FIRST_TOKEN_ABBREVIATIONS:
        tokens[0] = FIRST_TOKEN_ABBREVIATIONS[tokens[0]]
    tokens = [ABBREVIATIONS.get(t, t) for t in tokens]
    text = " ".join(tokens)

    # Strip trailing legal suffixes until none matches, then a leading THE.
    changed = True
    while changed:
        changed = False
        for suffix in LEGAL_SUFFIXES:
            if text.endswith(f" {suffix}"):
                text = text[: -(len(suffix) + 1)].strip()
                changed = True
                break
    if text.startswith("THE "):
        text = text[4:].strip()
    return text


def strip_level_suffix(recipient_id: str | None) -> str | None:
    """USAspending's recipient id without its ``-C``/``-R``/``-P`` level marker."""
    if not recipient_id:
        return None
    return _LEVEL_SUFFIX.sub("", recipient_id.strip()) or None


class HasIdentity(Protocol):
    """What identity resolution needs from a record. ``Award`` satisfies it."""

    recipient_uei: str | None
    recipient_id: str | None
    recipient_name: str | None


@dataclass(frozen=True)
class Identity:
    """How one record was resolved, and to what."""

    tier: str  # "uei" | "recipient_id" | "name"
    value: str

    @property
    def key(self) -> tuple[str, str]:
        return (self.tier, self.value)


def resolve(record: HasIdentity) -> Identity | None:
    """The strongest identity available for one record, or None if there is nothing."""
    uei = (record.recipient_uei or "").strip().upper()
    if len(uei) == 12:
        return Identity("uei", uei)
    stripped = strip_level_suffix(record.recipient_id)
    if stripped:
        return Identity("recipient_id", stripped)
    name = normalize_name(record.recipient_name)
    if name:
        return Identity("name", name)
    return None


def resolution_mix(identities: list[Identity | None]) -> dict[str, float]:
    """The share of records resolved at each tier, rounded to four places.

    Published in the report because it is the honest measure of how much the distinct-
    organization counts can bear.
    """
    total = len(identities)
    if not total:
        return {}
    counts = Counter(i.tier if i else "unresolved" for i in identities)
    return {tier: round(n / total, 4) for tier, n in counts.items()}


def name_tier_is_high(
    mix: dict[str, float], threshold: float = NAME_TIER_WARNING_THRESHOLD
) -> bool:
    """Whether enough records fell through to name matching to warrant a caveat."""
    return mix.get("name", 0.0) > threshold

File: awards/src/analysis/identity.ts
/**
 * Deciding when two award records are the same organization.
 *
 * A direct port of `precedent/analysis/identity.py`. The two statistics people quote from
 * this site - the new-entrant rate and the repeat-winner rate - are both counts of distinct
 * organizations, so splitting one grantee into two makes a program look more open than it
 * is and merging two makes it look closed. Either way the number reads as authoritative.
 *
 * This file and its Python original must agree exactly. Where a rule looks arbitrary it is
 * because it matches the reference implementation the numbers were verified against; change
 * one without the other and the site quietly disagrees with the tool it is a companion to.
 */

/** Applied repeatedly: "FOO SERVICES INC LLC" loses both, "THE FOO CO" loses CO then THE. */
const LEGAL_SUFFIXES = [
  "INCORPORATED",
  "CORPORATION",
  "COMPANY",
  "LIMITED",
  "INC",
  "LLC",
  "LLP",
  "LP",
  "CORP",
  "CO",
  "LTD",
  "PC",
  "PA",
  "THE",
];

/** Whole-token expansions only. Substring replacement would turn "COSTA" into "COMPANYSTA". */
const ABBREVIATIONS: Record<string, string> = {
  DEPT: "DEPARTMENT",
  UNIV: "UNIVERSITY",
  ASSN: "ASSOCIATION",
  ASSOC: "ASSOCIATION",
  NATL: "NATIONAL",
  INTL: "INTERNATIONAL",
  SVCS: "SERVICES",
  SVC: "SERVICES",
  CTR: "CENTER",
  CNTY: "COUNTY",
  US: "UNITED STATES",
  AAA: "AREA AGENCY ON AGING",
};

/** "ST LOUIS" is a city and "ST STATE" is nonsense, so this expands only in first position. */
const FIRST_TOKEN_ABBREVIATIONS: Record<string, string> = { ST: "STATE" };

/** The share resolved by name above which the statistics deserve a warning. */
export const NAME_TIER_WARNING_THRESHOLD = 0.1;

export type Tier = "uei" | "recipient_id" | "name";

export interface Identity {
  tier: Tier;
  value: string;
}

export interface HasIdentity {
  recipientUei: string | null;
  recipientId: string | null;
  recipientName: string | null;
}

export function identityKey(identity: Identity): string {
  return `${identity.tier}:${identity.value}`;
}

/** One organization name reduced to a comparable form. */
export function normalizeName(raw: string | null | undefined): string {
  if (!raw) return "";
  // NFKD then strip combining marks, matching the Python unicodedata pass.
  let text = raw
    .normalize("NFKD")
    .toUpperCase()
    .replace(/\p{M}/gu, "")
    .replace(/&/g, " AND ")
    .replace(/[^A-Z0-9 ]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
  if (!text) return "";

  const tokens = text.split(" ");
  const first = tokens[0];
  if (first !== undefined && FIRST_TOKEN_ABBREVIATIONS[first] !== undefined) {
    tokens[0] = FIRST_TOKEN_ABBREVIATIONS[first] as string;
  }
  text = tokens.map((t) => ABBREVIATIONS[t] ?? t).join(" ");

  // Strip trailing legal suffixes until none matches, then a leading THE.
  let changed = true;
  while (changed) {
    changed = false;
    for (const suffix of LEGAL_SUFFIXES) {
      if (text.endsWith(` ${suffix}`)) {
        text = text.slice(0, -(suffix.length + 1)).trim();
        changed = true;
      }
    }
  }
  if (text.startsWith("THE ")) text = text.slice(4).trim();
  return text;
}

/**
 * USAspending's recipient id without its `-C`/`-R`/`-P` level marker.
 *
 * The same organization appears as child, recipient and parent; keeping the suffix splits
 * one organization into three, and the reference implementation's verification notes name
 * this as one of the two most likely causes of a wrong answer.
 */
export function stripLevelSuffix(recipientId: string | null | undefined): string | null {
  if (!recipientId) return null;
  return recipientId.trim().replace(/-(C|R|P)$/i, "") || null;
}

/** The strongest identity available for one record, or null if there is nothing. */
export function resolve(record: HasIdentity): Identity | null {
  const uei = (record.recipientUei ?? "").trim().toUpperCase();
  if (uei.length === 12) return { tier: "uei", value: uei };
  const stripped = stripLevelSuffix(record.recipientId);
  if (stripped) return { tier: "recipient_id", value: stripped };
  const name = normalizeName(record.recipientName);
  if (name) return { tier: "name", value: name };
  return null;
}

/** The share of records resolved at each tier, rounded to four places. */
export function resolutionMix(identities: (Identity | null)[]): Record<string, number> {
  const total = identities.length;
  if (!total) return {};
  const counts = new Map<string, number>();
  for (const identity of identities) {
    const tier = identity ? identity.tier : "unresolved";
    counts.set(tier, (counts.get(tier) ?? 0) + 1);
  }
  const out: Record<string, number> = {};
  for (const [tier, n] of counts) out[tier] = Math.round((n / total) * 10000) / 10000;
  return out;
}

export function nameTierIsHigh(
  mix: Record<string, number>,
  threshold = NAME_TIER_WARNING_THRESHOLD,
): boolean {
  return (mix.name ?? 0) > threshold;
}
